File: codes/3.ProgrammingLanguage/编程语言试水/Score100WordsAnalyzer.swift
import Foundation

let start = Date().timeIntervalSinceReferenceDate
let fileData = try Data(contentsOf: URL(fileURLWithPath: "Dictionary.txt"))
let text = String(decoding: fileData, as: UTF8.self)
let lines: [String] = text.components(separatedBy: "\r\n")
let a = UnicodeScalar("a").value
let score100Words = lines
    .filter { !$0.contains("-") && !$0.contains(" ") && !$0.contains("=") }
    .filter { word -> Bool in
        var scoreSum = 0
        for c in word.unicodeScalars {
            if c >= "a" && c <= "z" {
                scoreSum += Int(c.value - a) + 1
            } else {
                return false
            }
        }
        return scoreSum == 100
    }
print(score100Words)
print("\nwords number: \(score100Words.count)")
let end = Date().timeIntervalSinceReferenceDate
print("        time: \((end - start) * 1000) ms")
/*
编码体验：
1、对 Windows 平台的支持非常不友好，在 Windows 10 上花了四个小时都没有跑通 Hello World
2、对于字符的处理很繁琐，比如得到其整型值需要 UnicodeScalar("a").value ，而 java 只需要 (int) 'a'
3、没有分号
4、没有main方法（可以有）
5、区分可变和不可变的变量（let、var）
6、自动推断变量类型
7、构造方法指定参数名称，如：String(decoding: fileData, as: UTF8.self)
*/

File: codes/3.ProgrammingLanguage/编程语言试水/Score100WordsAnalyzer.exs
start = System.system_time(:millisecond)
score100Words = File.stream!("Dictionary.txt")
  |> Enum.map(&String.trim(&1))
  |> Enum.filter(fn line ->
    if String.contains?(line, "-") || String.contains?(line, "=") || String.contains?(line, " ") do
      false
    else
      line |> String.to_charlist() |> Enum.map(&(&1 - 96)) |> Enum.sum()  == 100
    end
  end)
  |> Enum.to_list()
IO.puts Enum.join(score100Words, ", ")
IO.puts "\nwords number: " <> Integer.to_string(length(score100Words))
IO.puts "        time: " <> Integer.to_string(System.system_time(:millisecond) - start) <> "ms"
# 编码体验：
# 加分项:
# 1、不需要引入包
# 2、不需要定义同名类，如：public class Score100WordsAnalyzer
# 3、不需要声明 main 方法
# 4、不需要声明变量类型
# 5、变量前不需要 let 或 var
# 6、行尾不需要分号
# 7、整体来说代码算比较简洁的
# 减分项：
# 1、有些写法性能很差，比如：|> Enum.filter(&(!String.contains?(&1, "-") && !String.contains?(&1, " ") && !String.contains?(&1, "=")))
# 2、字符串与整数连接时，需要强转一下，不方便，比如："\nwords number: " <> Integer.to_string(length(score100Words))
# 3、基于变量的方法很少，比如我更习惯 line.contains("-") 而不是 String.contains?(line, "-")
